spec.html

```diff
@@ -28152,6 +28445,112 @@ <h1>
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-with-clause-to-attributes" type="sdo">
+        <h1>Static Semantics: WithClauseToAttributes ( ): a List of ImportAttribute Records</h1>
+        <dl class="header">
+        </dl>
+
+        <emu-grammar>
+          WithClause : `with` `{` `}`
+        </emu-grammar>
+        <emu-alg>
+          1. Return a new empty List.
+        </emu-alg>
+
+        <emu-grammar>
+          WithClause : `with` `{` WithEntries `,`? `}`
+        </emu-grammar>
+        <emu-alg>
+          1. Let _attributes_ be WithClauseToAttributes of |WithEntries|.
+          1. Sort _attributes_ according to the lexicographic order of their [[Key]] fields, treating the value of each such field as a sequence of UTF-16 code unit values. NOTE: This sorting is observable only in that hosts are prohibited from distinguishing among attributes by the order they occur in.
+          1. Return _attributes_.
+        </emu-alg>
+
+        <emu-grammar>WithEntries : AttributeKey `:` StringLiteral</emu-grammar>
+        <emu-alg>
+          1. Let _key_ be the AttributeName of |AttributeKey|.
+          1. Let _entry_ be the ImportAttribute Record { [[Key]]: _key_, [[Value]]: SV of |StringLiteral| }.
+          1. Return « _entry_ ».
+        </emu-alg>
+
+        <emu-grammar>WithEntries : AttributeKey `:` StringLiteral `,` WithEntries</emu-grammar>
+        <emu-alg>
+          1. Let _key_ be the AttributeName of |AttributeKey|.
+          1. Let _entry_ be the ImportAttribute Record { [[Key]]: _key_, [[Value]]: SV of |StringLiteral| }.
+          1. Let _rest_ be WithClauseToAttributes of |WithEntries|.
+          1. Return the list-concatenation of « _entry_ » and _rest_.
+        </emu-alg>
+
+        <emu-note>
+          <p>This section is extended by <emu-xref href="#sec-deprecated-assert-keyword-for-import-attributes"></emu-xref>.</p>
+        </emu-note>
+      </emu-clause>
+
+      <emu-clause id="sec-deprecated-assert-keyword-for-import-attributes" deprecated normative-optional>
+        <h1>Deprecated `assert` keyword for Import Attributes</h1>
+
+        <p>Import Attributes are denoted using the `with` keyword. However, there are existing implementations based on a previous version of the proposal using the `assert` keyword. Due to potential web compatibility risks, the `assert` keyword is still included in this specification, but its removal is being investigated.</p>
+
+        <p>This deprecated feature includes the `assert` fallback for import attributes in import calls (step <emu-xref href="#step-evaluate-import-call-assert-fallback"></emu-xref> of <emu-xref href="#sec-evaluate-import-call"></emu-xref>).</p>
+
+        <p>The following augments the |ImportDeclaration| production in <emu-xref href="#sec-imports"></emu-xref>:</p>
+        <emu-grammar type="definition">
+          ImportDeclaration :
+            `import` ImportClause FromClause [no LineTerminator here] AssertClause `;`
```

Comment:
This all works, but the intention is that `assert` behaves identically to `with`, right?

If so, you can accomplish this with much less duplication:

Define a new `WithKeyword` as
```
WithKeyword :
  `with`
```

and `WithClause` as

```
WithClause :
  WithKeyword `{` `}`
  WithKeyword `{` WithEntries `,`? `}`
```

And then the deprecated bit can just replace the definition of `WithKeyword`:
```
WithKeyword :
  `with`
  `assert`
```

That way you don't need to duplicate anything other than that one production.

Reply:
I plan to remove the `[no LineTerminator here]` restriction when using `with` (https://github.com/tc39/ecma262/pull/3057#discussion_r1187360998). It was mostly an oversight from when changing the keyword, we just forgot to remove it. I'll quickly bring it up at the next plenary.

However, this means that there will be a slight difference between the `assert` and `with` syntax, so we need the two separate productions.

Reply:
Ah, got it.